### DISCUSSION TITLE
HasMany with Attach

### DISCUSSION THREAD

**Discussion by pxlrbt** [laravel/nova-issues#3191] [Ideas & Feature Requests]
Moving the discussion from https://github.com/laravel/nova-issues/issues/520 to here

**Comment by schonhoff** [comment]
As @crynobone asked. Here is my flow:

Company Z:
- Car A
- Car B
- Car C
- Car D

Company Y:
 - Car E
 - Car F

At the end of the year Company Y get 2 Cars from Company Z. The company got Car A and Car C.

In Nova I need to go to the cars and remove them from the Company Z and attach them to the Company Y. If I go to the company resource I only can create new cars but can't attach them to the resource and I can deattach them. With only 2 Cars it is quite simple but a company may have 30 cars switching at the end of a year it is painful.

Is it more understandable @crynobone ?

**Reply by schonhoff** [reply]
Another example from that thread:

Company has many users and a user belongs to one company.
If I want to attach a user to a company I need to go into the user resource and attach the user to the company on this way. The other way around doesn't work because the company resource only gives us the option to create new users.

Maybe a solution for this problem would be:

Requirements:
 - The user model has a company_id field.

Solution:
 - If there is a model existing with a non-set company_id it can be attach to a company. 

This would be possible with an "exists" request on the foreign key column of the related model. If the exists request is true the button of attaching users to the company can be visible and the list of possible users, which can be attach to the company, can be all "null" users.

Is that maybe a possible solution?

**Comment by pxlrbt** [comment]
@crynobone 

I think this is mainly useful for quickly editing data. In my example there where some locations. Each location `hasMany` events and every event `belongsTo` a location. At the time creating the events, the location of these events hasn't been determined yet. To fill the site with data we create e.g. 10 locations and 100 events.

One week later the client has decided which events will be hosted in which location. He has a list with all 10 locations and the events for each location. As the list is sorted for locations he would want to edit a location. But he can only create new events from there and not attach existing ones.

**The current workflow would be:**
- Go to events resource
- Open first event
- Check which location this event should have
- Set location
- Save event
- Proceed with event 2
- ...
- Proceed with event 100

**The ideal workflow would be:** 
- Go to location resource
- Open first location
- Check which events are hosted at this location
- Attach event 1 - 10
- Proceed with location 2
- ...
- Proceed with location 10

**Reply by schonhoff** [reply]
This is what I tried to say. Thank you @pxlrbt . I'm not a native English speaker so it is quite difficult for me to describe the workflow. But that would be my use-case aswell.

**Comment by pxlrbt** [comment]
Regarding https://github.com/dillingham/nova-attach-many: The readme states that it's a `BelongsToMany` form field. We are talking about `HasMany` relations.

Maybe this can be a workaround for this issue. **BUT:** It's added to the edit form and not on the detail page where the `HasMany` field is added. So the UI for creating/attaching resources is split up on two pages. That's not ideal and probably hard to grasp for many clients that just expect to attach a resource where they would normally do as with the `BelongsToMany`

**Comment by crynobone** [comment]
Is it correct for me to assume all your requirements set belongs to foreign id as `nullable`?

```php
$table->foreignId($parentId)->nullable();
```

**Reply by pxlrbt** [reply]
Can't find a piece of code right now, as I used `belongsToMany` as a workaround everywhere. But I think that's right. Otherwise you couldn't create resources without the relation.

**Reply by schonhoff** [reply]
In my case that is correct. In my example a vehicle can be submitted by a fleet manager and can be related by a company owner. That is why it is nullable.

**Reply by borisson** [reply]
That is how this works for us as well, the relation has to be nullable for the "unset" functionality to work.

**Reply by mjbnz** [reply]
> Can't find a piece of code right now, as I used `belongsToMany` as a workaround everywhere. But I think that's right. Otherwise you couldn't create resources without the relation.

Could you elaborate on how you use `belongsToMany()` in this scenario? I've tried it, and it does work, but every detail view throws the error:
```
Call to undefined method Illuminate\Database\Eloquent\Relations\HasMany::getPivotClass()
```

I can definitely confirm however that the specific scenario that this would be useful, is when:
- The `HasMany()` related Model has a foreign key referencing the `id` on the Resource.
- The foreign key is nullable. This is a hard requirement, or it's a pointless feature.

Some thoughts about implementation:
- Attach would set that foreign key to the current Resource id.
- Detach would set the foreign key to `NULL`.
- Attaching would ideally have a `relatableQuery()` that limits the results to Models with a `NULL` in the foreign key field.
- However, it could be useful to allow the transfer of the Model between resources, so that's not a requirement... that would be up to the user.
- Attaching could realistically allow multiple attachments at once, but that's probably beyond the scope of this issue.

**Reply by mjbnz** [reply]
> > Can't find a piece of code right now, as I used `belongsToMany` as a workaround everywhere. But I think that's right. Otherwise you couldn't create resources without the relation.
> 
> Could you elaborate on how you use `belongsToMany()` in this scenario?

@pxlrbt I've now realised you're probably standing up an entire pivot table relationship to maintain the One to Many links... is that what you're doing?

**Reply by pxlrbt** [reply]
> @pxlrbt I've now realised you're probably standing up an entire pivot table relationship to maintain the One to Many links... is that what you're doing?

Yes exactly. You change the whole relation and therefore need that additional pivot table. That messes up the database structure and is especially annoying when you already have data that needs to be migrated as well. But at least it works (as a hack).

**Reply by mjbnz** [reply]
> Yes exactly. You change the whole relation and therefore need that additional pivot table. That messes up the database structure and is especially annoying when you already have data that needs to be migrated as well. But at least it works (as a hack).

Ahh... definitely a less than ideal solution. Thanks for your feedback.

@crynobone what are your thoughts on this?

**Comment by flatcapco** [comment]
I still agree this would be a really useful feature too.

Imagine if you had a Quote and wanted to attach multiple existing "availabilities" in my case I have about 10,000 

At no point would the admin user ever be able to create those availabilities as they are handled by the system.


At the moment I'm using an action to attach many.. which is fine but then displaying them the interface isn't as good as the default HasMany table... or anywhere near as useful.

I'm not sure why this feature has taken so long so please let me know if we're still missing any crucial information to push it over into a version?

**Comment by coatesjonathan** [comment]
Hello,

I have this issue too, I use the `HasManyThrough` a lot in my current project, as each User needs to have many `Specialisms`, so I created a model called `Specialism` and another model called `UserSpecialism` which the user model then accesses the many `Specialisms` through there `UserSpecialism` table, using a HasManyThrough relationship...

Having a way to attach existing specialisms to existing users would be awesome, and not having to create new ones all the time because Nova doesn't allow the `Attach` method.

**Comment by gepopp** [comment]
Wow...

@taylorotwell slowly but surly i think buying a v4 license was a shot in the foot. besides some other issues, i allready posted here, this also still is not solved, what sense does it make if i cannot attach existing models on a has many relation.... mby u should have a look to a wordPress backend this is made for non programming end users

**Comment by arjankapteijn** [comment]
Same issue here; I have  list of activities and simple want to attach them to a project. A project can have multiple activities, a activity can only be attached to one project. 

Project.php
HasMany::make('Activities'),

Activity.php
BelongsTo::make('Project')

I can now edit a activity to connect it to a project, but when I view a project I can only create a new activity instead of attach it to an existing one.

**Comment by schonhoff** [comment]
@crynobone Is there a possibility for this to be added in Nova 4? Maybe with another Button next to the Attach button that could be called "Attach existing xyz"?

**Comment by bendadaniel** [comment]
I dont understand why this basic function, which I need to use on every second resource is not in Laravel Nova. See other discussions and issues, this basic "Attach existing model" feature want hundreds of people...

**Comment by reppair** [comment]
OK, I was wondering how to make it happen, so basically after reading the [closed issue](https://github.com/laravel/nova-issues/issues/520) (without explanation) I ended up here. This is a common case.

@taylorotwell care to let us know what do you think about that request?

**Reply by reppair** [reply]
On second thought, I think it does not make much sense to "attach" another existing item, basically by doing so it will also be "detached" from another model (unless the foreign key is nullable). Therefore, this would be rare case, better go to the related modal resource and update the belonging model.
